### DISCUSSION TITLE
Mutation version of `queryOptions()`.

### DISCUSSION THREAD

**Discussion by nix6839** [TanStack/query#6096] [Ideas] [answered]
It would be nice to have a mutation version of `queryOptions()`.

**Comment by TkDodo** [comment, marked as answer]
what would you need it for? The main reason for `queryOptions` is to have a way to share options between `useQuery`, `useQueries` and things like `queryClient.prefetchQuery`. There's no equivalent for mutations where this would be needed

**Reply by nix6839** [reply]
Just type guarding.
I manage all my queries and mutations as objects, because it's easier to reuse and define.
It would be nice to be able to use type guarding for this.

**Reply by chrisirhc** [reply]
In my use case, I was trying to make sure that all the mutations use the same mutationCache/options similar to how useMutation behaves via the hook.
Is there no equivalent because mutations don't typically share defaults?

**Reply by TkDodo** [reply]
`queryClient.setMutationDefaults(mutationKey, options)` is what you want, or, if you want it globally for all mutations, you can use the defaultOptions on the queryClient itself

**Reply by chrisirhc** [reply]
Hm there’s not getMutationDefaults? Or am I missing something. Unless I use the hooks (which I can’t, in my use case for class components), I can’t really use queryClient.doMutation to get mutation action function with the options + cache setup correctly. 

And fetchQuery uses a different set of options (for queries).

**Reply by TkDodo** [reply]
https://github.com/TanStack/query/blob/b67107114ce6ccf0d38c2ab22b3088df19ec119b/packages/query-core/src/queryClient.ts#L425-L449

**Reply by chrisirhc** [reply]
Ah I my bad, I missed that. update: Actually, it wasn’t that I missed that. I didn’t wasn’t sure you could do that and whether it’s right use of the method. 
So for mutations, I’d do `queryClient.fetchQuery({…queryClient.getMutationDefaults(mutationKey), …otherOpts})` ?

**Reply by TkDodo** [reply]
No, you set the options once as the default and then they'll be automatically picked up

**Reply by chrisirhc** [reply]
I’m still confused as to how it would be picked up since there’s no `queryClient.mutate(mutationKey)` method, I don’t see how I can make use of the mutation cache or the default mutation options given a queryClient instance.
Is this feature only available through hooks?

**Reply by TkDodo** [reply]
okay now I'm confused. Why does a query need to know the default options of a mutation? They have different types and things ...

sorry, I'm done discussing this unless you show a reproduction of what you want to achieve. I need to remind myself that there is a reason why I always require a runnable reproduction upfront - exactly to avoid confusions like this.

**Reply by chrisirhc** [reply]
No worries. After reading `useMutation` implementation, I realized that my questions were related to not knowing about the `MutationObserver`. 

I was trying to initiate mutations with only the queryClient but instead I needed to create a `MutationObserver` as is done in `useMutation`. 

Initially, I thought of mutation as a "dual" (symmetrical) action as query but instead that's not the right mental modal. However, instead mutations are built on top of MutationObserver that know about queries and queryClient, and they're not "part of" queryClient.

Cheers!

**Reply by TkDodo** [reply]
you can execute a mutation from the queryClient like so (we do this in tests):

https://github.com/TanStack/query/blob/5d29f8ff47958a2349fdc4ca682c3630beeabc9d/packages/query-core/src/tests/utils.ts#L35-L44

**Reply by chrisirhc** [reply]
That's what I needed. Thank you! 

I thought that's what the original author was asking for but perhaps I'm wrong. This answers the question I had.

**Reply by dhulme** [reply]
We use `queryOptions` to help calling the same query in different components, and keep the API calling logic separate to the component. A `mutationOptions` function would allow us to do this same sharing/separation for mutations. Do you have any thoughts on this @TkDodo?

**Reply by TkDodo** [reply]
I don't think it's really necessary because you can just write custom mutation hooks? The advantage of `queryOptions` is that you can share them between hooks and imperative methods, like `prefetchQuery` in a type-safe way. At runtime, it doesn't do anything.

> and keep the API calling logic separate to the component.

imo that's what a request layer, let's call it `myApi`, is for. It can abstract axios or fetch in a type-safe way so that you can call them in hooks:

```
const useUpdateUser = () => {
  const queryClient = useQueryClient()
 
  return useMutation({
    mutationFn: myApi.updateUser,
    onSuccess: (_, id) => {
      return queryClient.invalidateQueries(userQuery(id))
    }
  })
}
```

What's the advantage of extracting that logic to outside of hooks? How would you get access to the queryClient outside of hooks?

**Reply by dhulme** [reply]
Thank you for your suggestions. We had custom hooks for queries/mutations before v5, but the section in the docs on the new `queryOptions` led us to start moving away from custom hooks, but I see now they can both be used together.

I think it would be really useful to have a section in the docs of recommended practices for making queries/mutations reusable?

**Comment by smashercosmo** [comment]
I have another use case for `mutationOptions`. This will make it easier to create custom hooks. WDYT @TkDodo ?
```ts
function someMutationOptions() {
  return mutationOptions({
    mutationFn: ...
  });
}

type AdditionalOptions = Omit<
  ReturnType<typeof someMutationOptions>,
  "mutationFn"
>;

export default function useSomeMutation(
  additionalOptions?: AdditionalOptions
) {
  const options = someMutationOptions();

  return useMutation({
    ...options,
    ...additionalOptions,
  });
}
```

In this case additionalOptions type will be inferred from the mutationFn and we won't have to type options object manually

**Reply by TkDodo** [reply]
```
import { type MutationOptions } from '@tanstack/react-query'

function someMutationOptions() {
  return {
    mutationFn: ...
  } satisfies MutationOptions;
}
```

**Reply by smashercosmo** [reply]
I'm not sure, I understand your solution. MutationOptions will require providing generics values. And I want all typings to be inferred from mutationFn.

**Reply by TkDodo** [reply]
this works:

https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzjAnmApnAsgVxgQxmAgDsB5MI0gZwBo4dr1cCqS4BfOAMyghDgByAAIES1AgGMA1gHoo6fJJgBaAI450UFIIBQu7jhLLi7av2Z5Cpim2oAKAJSJdcOApg4o7BK7dwQKzYAMRIALjh7YBIwPAiSHBAAIy1nAF4APjgABT4QYCYAOgVzABsAN3QomLxCmAgAZRgoaIBzJ0c-Lmpram5gdGpsIJtKU2oAHnwSFHpp2bh5uZmMgG5dDn1JGngkQNZMLjSGJhZrUntzEEtWUbsOvzdZWTcAPQB+fX3CKoBGRyAA

**Reply by smashercosmo** [reply]
Yeah, but this doesn't solve the original problem with custom hook and additional options :)
https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzjAnmApnAsgVxgQxmAgDsB5MI0gZwBo4dr1cCqS4BfOAMyghDgByAAIES1AgGMA1gHoo6fJJgBaAI450UFIIBQu7jhLLi7av2Z5Cpim2oAKAJSJdcOApg4o7BK7dwQKzYAMRIALjh7YBIwPAiSHBAAIy1nAF4APjgABT4QYCYAOgVzABsAN3QomLxCmAgAZRgoaIBzJ0c-Lmpram5gdGpsIJtKU2oAHnwSFHpp2bh5uZmMgG5dDn1UDDgAQQATfeA2fFLbcbg0uAAldE9vABU0dAnt9AhuOHMQS1ZRuwy+kMxjYDCYDQsLGspHs+EOx1Mp3ONAiByOJzOYxozl8bkkNHgECx4kuX0hI1IyPETj8Hi87EYv2hJHsuP8hQ5RLstD8bg5hTh6MRmO5XU6m10jIhPyhbFZflIDRwkkkgwcOM2HEcQA

**Reply by TkDodo** [reply]
> with custom hook and additional options

and neither does `queryOptions`. The take is that with an options abstractions, you don't need a custom hook anymore that can take additional options, because you can just spread them on the usual call

**Reply by smashercosmo** [reply]
Well, but it does https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgRwK4FMoE8DyYbAQB2AzgDRyonoCKG2cAvnAGZQQhwDkAAjAIakBAYwDWAeijp+wmAFo0mLFwBQKlqiKzCROCQ616ufDpIAKYETCoYALjhFUIAEaYAlIhVw4UmKii6ith4BMTmCF7eKEYA0uhY9gDaXAAW6AA26RBcFJbWMAC6ZJHeQVgAYkT2Zh4AvAB8nlFRvv66AArsIMDUAHRS+ukAbugWVja9MBAAyjBQlgDmNW4lTJGMK4xqMFhg6HAAggAmR8ChRPzpIaZwtXA43TAAPJEASuh+AQAqu+hPO3sICw9AY6EprmF6sVvAAiMpxLAwuAAHzgcKMlRhKnqAG41Cp0AAPSCwOBHdAsfiodLwDRac6UajTUFGMyRPI2eyOFyYaFwfgnM46S4Q0gAfnsx1O5xFJjCKg8EW8wjC8AgctItxBIEM4I15g5MDceMirQCjN12DMSqivTt6vO5FWdt6Aulwqu+r5GzxWxUVHQzJ1YKtAEYAAwUG3UdLoWTVI78AR1RqJgQqDZAA

You can see that 'select' function argument is inferred from the queryFn here and I don't need to type it manually

**Reply by TkDodo** [reply]
no, it doesn't really work, because you can't transform data in select. The signature is fixed to (string) => string. If you try to return anything but a string, it won't work: [see this reproduction](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgRwK4FMoE8DyYbAQB2AzgDRyonoCKG2cAvnAGZQQhwDkAAjAIakBAYwDWAeijp+wmAFo0mLFwBQKlqiKzCROCQ616ufDpIAKYETCoYALjhFUIAEaYAlIhVw4UmKii6ith4BMTmCF7eKEYA0uhY9gDaXAAW6AA26RBcFJbWMAC6ZJHeQVgAYkT2Zh4AvAB8nlFRvv66AArsIMDUAHRS+ukAbugWVja9MBAAyjBQlgDmNW4lTJGMK4xqMFhg6HAAggAmR8ChRPzpIaZwtXA43TAAPJEASuh+AQAqu+hPO3sICw9AY6EprmF6sVvAAiMpxLAwuAAHzgcKMlRhKnqAG41Cp0AAPSCwOBHdAsfiodLwDRac6UajTUFGMyRPI2eyOFyYaFwfgnM46S4Q0gAfnsx1O5xFJjCKg8EW8wjC8AgctItxBIEM4I15g5MDceMirQCjN12DMSqivTt6vO5FWdt6Aulwqu+r5GzxWxUKqEiDJ-AETC1VHQzJ1YKtAEYAAwUG3UdLoWTVQl1RoAOScrigZkzKg2QA).

The idea is really that with options, you don't need a custom hook that accepts dynamic options. Why would I pass `select` to `useSomeQuery` if I could also just do:

```
const { data } = useQuery({
  ...someQueryOptions(10),
  select: (x) => Number(x)
})
```

which will work perfectly fine ...

**Reply by smashercosmo** [reply]
I get the idea, but if I remember correctly the general recommendation by react-query team was that you should create custom hooks which increases code readability. So instead of
```
 const { data } = useQuery({
  ...getTodosQueryOptions(),
  ...additionalOptions
})
```

you should have
```
 const { data } = useGetTodosQuery(additionalOptions)
```

**Reply by TkDodo** [reply]
that recommendation has likely changed with `queryOptions`. I mean, you can still make custom hooks like before, but the recommendation about building custom hooks that accept arbitrary generic options was always: don't do it. So I think this is fine:

```
const useGetTodosQuery: ({ enabled }: { enabled: boolean}) => {
  return useQuery({ ...getTodosQueryOptions(), enabled })
}
```

even though I would also probably inline it. But having a whole `options` object be typed with `QueryOptions` and 4 type parameters is just not worth it and never was.

**Reply by hossamAlhamidi** [reply]
what if i wanted to write a global mutation hook that does exactly what useMutation do , it serves as a wrapper so if i wanted to migrate to a new version for example it would be easier , write now , im gonna lose auto complete functionality. 
with queryOptions replacing UseQueryOptoins i can still do that 
`  const { data, isLoading } = useGlobalQuery({
    queryKey: ["todos"],
    queryFn: () => axiosInstance.get("https://jsonplaceholder.typicode.com/todos"),
    ...queryOptions({
      staleTime: 1000 * 60 * 5
    })
  });`
  but on mutation i can't, so i think i can be helpful

**Reply by junwen-k** [reply]
Hi @TkDodo, I’d like to clarify a few things and get your thoughts on the following:

1. Is the current recommendation to use `queryOptions` directly with `useQuery`, `useSuspenseQuery`, etc. the recommended approach now, with `useXQuery` hooks no longer needed / encouraged?

2. What if I need a custom `useXMutation` hook with built-in `invalidateQueries`, while still allowing users to override specific options like `onSuccess` to display a toast, for example? Would you recommend manually typing the `onSuccess` callback? This seems like a common scenario, as seen in [bulletproof-react’s delete comment example](https://github.com/alan2207/bulletproof-react/blob/master/apps/nextjs-pages/src/features/comments/components/delete-comment.tsx#L19). Similarly, this [example](https://github.com/alan2207/bulletproof-react/blob/master/apps/nextjs-app/src/features/discussions/api/get-discussions.ts#L34) demonstrates `useDiscussions` with generic options, which I understand is something you generally advise against. Could you share your thoughts on this approach?

It looks like this is all that's needed (taken from bulletproof-react) to support a fully type-safe and flexible hook with `queryConfig` and `mutationConfig` overrides:

```ts
// src/lib/react-query.ts
export type ApiFnReturnType<FnType extends (...args: any) => Promise<any>> =
  Awaited<ReturnType<FnType>>;

export type QueryConfig<T extends (...args: any[]) => any> = Omit<
  ReturnType<T>,
  'queryKey' | 'queryFn'
>;

export type MutationConfig<
  MutationFnType extends (...args: any) => Promise<any>,
> = UseMutationOptions<
  ApiFnReturnType<MutationFnType>,
  Error,
  Parameters<MutationFnType>[0]
>;

// src/features/discussions/apis/update-discussion.ts
type UseUpdateDiscussionOptions = {
  mutationConfig?: MutationConfig<typeof updateDiscussion>;
};

export const useUpdateDiscussion = ({
  mutationConfig,
}: UseUpdateDiscussionOptions = {}) => {
  const queryClient = useQueryClient();

  const { onSuccess, ...restConfig } = mutationConfig || {};

  return useMutation({
    onSuccess: (data, ...args) => {
      queryClient.refetchQueries({
        queryKey: getDiscussionQueryOptions(data.id).queryKey,
      });
      onSuccess?.(data, ...args);
    },
    ...restConfig,
    mutationFn: updateDiscussion,
  });
};

// src/features/discussions/apis/get-discussion.ts
export const getDiscussionQueryOptions = (discussionId: string) => {
  return queryOptions({
    queryKey: ['discussions', discussionId],
    queryFn: () => getDiscussion({ discussionId }),
  });
};

type UseDiscussionOptions = {
  discussionId: string;
  queryConfig?: QueryConfig<typeof getDiscussionQueryOptions>;
};

export const useDiscussion = ({
  discussionId,
  queryConfig,
}: UseDiscussionOptions) => {
  return useQuery({
    ...getDiscussionQueryOptions(discussionId),
    ...queryConfig,
  });
};
```

3. I’ve seen recommendations against fully type-safe, generic hooks multiple times, like [here](https://github.com/TanStack/query/discussions/6096#discussioncomment-9479095) and [here](https://github.com/TkDodo/blog-comments/discussions/102#discussioncomment-9405518). Is the main concern the complexity of getting generics typed correctly? I can see how this might not be worth the effort in many cases if not all options or flexibility are needed, but if the complexity can be managed easily, are there any other downsides?

Hi @smashercosmo, I’ve also created a TypeScript playground [here](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBDAnmApnA3nAIigZgQwFcAbGAUSimjgF85cqQ4AiAARnwDsBnDgYwDWAegCOhFFEQBaPtBTMA3AChQkWAmRpMAVW4oAsoQ4xgETgHkwJs91r1GLdl175BQqClcwpYiYkUq4NDwmIR6AIrikgA0cGEGRvjWnLHxkX4AwsTAKJwwsb6Slsm2dAwQTADkTjz8wh5ePlGIlUpKQkJw3FB8QtkARu6efN6FiAB0MNxtHXAAQvh6ACZwZnAAFjAwYNwAXB0A5sAw64T947IgQgAqXADKdaLNQkvA3Hxh3KY8QgBsAAwATl+AGJXu9Pt9LiBct5gQAOACsAEZ-gAmcZKFAADzU8FwhE4I2+cBAiWSxW+3AAPEoAJDXLBJfBwAC8cU4Ak4EAA7il6dcKFQoGzsHgiKQhdBogKAGr4KDAfD9YgoWzsgBuEGASxlDIyZhgOPg7MJXN5-IAfAAKekQKxU3ZwXQJYzfSk2amM5mxQWUaVwa7yxXK1XcX0GvLGy0ygCUTpdhjdZg9PC9TI4vqlUF9waVKrVEcN0Yw9I8MEIUE4qwdNmUNBmnW6vUKOW4QkgvG4k2mwCjUAIfDQAAUILxS3A4DqnbxFZwDspJyYYKqZzA5wulA2sbjgnBZLU4AcUDBR130kVazxRdbY2zLXAxqnuNb0EpJ2MANIoRBOgDalSdlMlQALoyh+zQAGKcE6t73nAf4gXAixwGeMCITKNCxo2zrLPQ1BIKgUh9rgEi5EOqycMQiCxOs5FoNy8A4niKBLJiB7jnwDRGmhSZJO6V7qqS5ICSUr7vsJyacNBsHLquXTrn2Bx3qyD7DowbwoOMHjcBAxAaigr5TksTqVJUsRyWgdAoWh2FYW0zF7hx8DxBk3EoLxIlrOyRlmHchB8EO3DhnA4xhfaJS0E6hEoBAuD7u5nlSc+KkPm+k7OY+zRZDkeSimk2XZLCt7KBJ5aVtW8R8ck4mTpOYUXIlY4wNVolUuBdV+QFQV7HA1oNQqBzcKlE51RBmRFXk4x9hq+DZEsSQoBebbWsep7NdwF6IM+t6xqVY2TrM1zrG8U62ASxDURoqC6ke3GYgdcCzNaYAKvgMJGlAd5dYFaq9da1oLRwTpobEs0hgWvWcIQID9BIsQHka2IwE6ZrcnyI3qRUmnUmjFoPgAPhy5oY3AROEkseB9qxEljeVVaUf5v3BQA-OM-VhYNw2LnVNAdfV4WCR19nbrMzZCPgYBgB2+DHj22JtJlmBksYVn5XobmeDxzWtWYtVPZ0ACS+5cHAYBUBqOpoFwiCrCcEg1pFdEeBZmhSNw+CkchnArLgqrYsABaYhJz2vVA70nhId5A-gIPNRJP09bBMcjeldWzMbfCm1AhIIBAizwM7KCxFwKwAAYzXNOqLctapl2dLD9IQwCkMRnDMHA-S25TBAkPAPJ0dWxvxHAZeuU1vC65wZcPbQmHYe0nQLHhaybNseyHMcpznJcNwCFgEBLBAQgqhABwyBUMJ5O24IfMFVJCKiaJgm8d9fGY0KwlIgIACz-IiKJ4SYgKn4fWDU1poU2s0HasYZQhyNibas5sICW0pt7W2EB7YigilSDYZFXZEQ9l7Uu9B-aB1VMHQ6nQXpUFQLARAd4+zHCVMQDMsc+poUQmTOAhtODMLmmwyChJiRmGpJwkChMOQ92pksO8AAyDhzUuFEwBpjJRIFYzYUnEwkwAjmT-hAluBeswGpKCAA) which solves what you are trying to do initially. It shows how we can maintain type-safe defaults while allowing overrides like `onSuccess`.

Thank you for all the work you’ve put into this library!

**Reply by TkDodo** [reply]
> Is the current recommendation to use queryOptions directly with useQuery, useSuspenseQuery, etc. the recommended approach now, with useXQuery hooks no longer needed / encouraged?

make `queryOptions` abstractions, and as long as your hook is just:

```
const useTodos = () => useQuery(todosOptions)
```

I don’t see a reason to make that extra abstraction. But sure, you can do it.

> What if I need a custom useXMutation hook with built-in invalidateQueries, while still allowing users to override specific options like onSuccess to display a toast, for example?

Not necessary as showing toast messages should likely to go the callbacks of `mutate`. But sure you can just make your custom hook accept: `onSuccess: () => void`, which should be enough for most cases.

**Comment by DylanVann** [comment]
I'm working on an e-commerce app where we want to show a state for mutation in progress on a certain item in a users cart and using `mutationOptions` to narrow the type of `options` here could be useful:

```tsx
import { updateItemQuantityMutationKey } from '@/shopify/updateItemQuantityMutationKey'
import { useQueryClient } from '@tanstack/react-query'

export const useIsModifyingLineItem = (id: string) => {
  const queryClient = useQueryClient()
  return (
    queryClient.isMutating({
      mutationKey: updateItemQuantityMutationKey,
      predicate: (options) => {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        return Boolean(options.state.variables?.lines?.find((line: any) => line?.id === id))
      },
    }) > 0
  )
}
```

Although I think in this case I'm just going to hoist the state up slightly so I don't have to use this code. If I was able to narrow the type using something like `mutationOptions` I might have left it.

**Comment by andyto** [comment]
Tried the following and seems to work well for me.
```ts
import type { DefaultError } from "@tanstack/query-core";
import type { UseMutationOptions } from "@tanstack/react-query";

export function mutationOptions<
  TData = unknown,
  TError = DefaultError,
  TVariables = void,
  TContext = unknown,
>(
  options: UseMutationOptions<TData, TError, TVariables, TContext>,
): UseMutationOptions<TData, TError, TVariables, TContext> {
  return options;
}
```

**Reply by leognmotta** [reply]
thanks

**Reply by iicdii** [reply]
You can actually import `DefaultError` from `@tanstack/react-query` directly instead of `query-core` since `react-query` re-export [core](https://github.com/TanStack/query/blob/bb25d069801694ffef66e67ac54da51b84ad680c/packages/react-query/src/index.ts#L4) so that you don't need to add `@tanstack/query-core` as a dependency in your project.


```diff
- import type { DefaultError } from "@tanstack/query-core";
- import type { UseMutationOptions } from "@tanstack/react-query";
+ import type { DefaultError, UseMutationOptions } from "@tanstack/react-query";
```

Final code:

```ts
import type { DefaultError, UseMutationOptions } from "@tanstack/react-query";

export function mutationOptions<
  TData = unknown,
  TError = DefaultError,
  TVariables = void,
  TContext = unknown,
>(
  options: UseMutationOptions<TData, TError, TVariables, TContext>,
): UseMutationOptions<TData, TError, TVariables, TContext> {
  return options;
}
```

**Comment by 0xbuidler** [comment]
Could we have an official implementation of `queryOptions` ? What's the blocker here ?

**Reply by TkDodo** [reply]
The “blocker” is still that I think it’s not needed. We added it for queries because the options need to be shared between various hooks and imperative functions. We can also leverage typings between queryKey and queryFn to make methods on the queryClient more type-safe. None of this is the case for mutations. There’s no reason to not just write a custom hook, and making this method available suggests otherwise.

**Comment by officer-kd6-3dot7** [comment]
> what would you need it for? The main reason for `queryOptions` is to have a way to share options between `useQuery`, `useQueries` and things like `queryClient.prefetchQuery`. There's no equivalent for mutations where this would be needed

I'd like this for `useMutationState`. I believe this hook currently needs the types to be manually passed.

**Reply by TkDodo** [reply]
Please show as TypeScript playground. I’d like to see where you need to pass type params to `useMutationState`.

**Reply by officer-kd6-3dot7** [reply]
👋 Hello TkDodo!! Thank you for responding. Here's my implementation of the `useMutationState` hook.

https://stackblitz.com/edit/vitejs-vite-ltocdd8d?file=src%2FTodos%2Findex.tsx

I’m still unsure whether `useMutationState` or `queryCache` manipulation is the better approach for handling optimistic updates. Is `useMutationState` recommended for managing optimistic updates?

Regarding passing types to the `useMutationState` hook, the generic doesn't seem to yield a helpful result. Type-casting the `variables` from the `select` callback seems to be the only way out in the my case.

Anyway, I imagined if there did exist a `mutationOptions` function, it would behave similar to `queryOptions`, eliminating the need for type-casting within mutation related APIs. What are your thoughts on this?

**Reply by TkDodo** [reply]
> eliminating the need for type-casting within mutation related APIs

this isn’t quite true. `useMutationState` can capture different mutations - the `mutationKey` is just a filter that isn’t required. You can say: give me all variables of all mutations that ever happened. There’s no way to type this automatically. Also, even if there is a `mutationKey` present on `mutationOptions` (which wouldn’t be guaranteed as it isn’t required), the `filter` you can pass to `useMutationState` matches mutations fuzzily. So with one key, you could match mutations of different shapes with different types of variables. That’s fundamentally different from `queryOptions`. For the same reason, we haven’t typed `queryClient.getQueryCache.find(filters)`, because everything there is fuzzy. No way to guarantee that a specific shape will be returned.

Further, just the presence of a `mutationOptions` function would not magically “solve” this issue. We would need to tie the `mutationKey` to `useMutationState` via something like the `DataTag` symbol we do on `queryOptions`.

**Reply by officer-kd6-3dot7** [reply]
I'm not saying that we "have to" tie this to `useMutationState`, but rather consider something that does "Exact" `mutationKey` matching, guaranteeing the shape of the `variables`. I would like to be able to observe mutations in a type-safe manner for a specific `mutationKey`. Maybe a hook or a method consuming the `dataTagSymbol` information, which provides access to all mutations that are exactly tied to that `mutationKey` and further provide filtering capabilities as options, using the inferred type. 

Optional `mutationKey` does pose a challenge to this idea, and I agree that calling it `mutationOptions` might not be entirely accurate. However, I still believe it’s worth exploring how we can achieve this in a type-safe way, especially since something similar has already been done with `queryKey` and `dataTagSymbol`.

**Comment by shoooe** [comment]
I understand that it's not strictly needed but on the other hand I don't understand what the downsides of providing such a function are.

"Make an API that is difficult to use incorrectly" is a great guideline for designing API but only when there are dangers to misusing the API. What are the dangers of creating a mutationOption and passing it to useMutation instead of creating a custom hook directly?

This also doesn't seem that costly in terms of API surface that needs to be added and people clearly would expect it to exist in a symmetrical way to queryOptions.

Just my 2 cents. As always: thanks for your contributions to this great library.

Peace ✌️

**Comment by TkDodo** [comment]
okay someone please make a PR, just to end this discussion 😂

**Reply by onyedikachi23** [reply]
I'd love to, but I have no experience contributing to OSS projects.

Rooting for anyone who can help ship this or help me contribute 🚀.

**Comment by Romej** [comment]
for anyone wondering, #8960 added it

**Reply by officer-kd6-3dot7** [reply]
Oh my god!!! Finally!!
